### PR TITLE
fix(nuget): don't fail when updating lockfiles with private registries

### LINKED ISSUE #7769: Error updating NuGet lock file when credentials are defined for a feed
**What Renovate type, platform and version are you using?**

Self-hosted v23.84.6 running in Bitbucket Pipelines.

**Describe the bug**

Renovate always fails to update lock file when a private feed requiring credentials is defined in nuget.config. The problem is pretty obvious from the logs - the command `dotnet nuget update source` is run with empty string as the value of username and password, but this command is not supported and throws an error.

**Relevant debug logs**

_Note: I redacted some of the data._

```
DEBUG: dotnet command (repository=**redacted**, branch=**redacted**)
       "cmd": [
         "dotnet nuget update source **redacted** --username **redacted** --password **redacted** --store-password-in-clear-text",
         "dotnet restore **redacted** --force-evaluate",
         "dotnet nuget update source **redacted** --username '' --password '' --store-password-in-clear-text"
       ]
DEBUG: Executing command (repository=**redacted**, branch=**redacted**)
       "command": "dotnet nuget update source **redacted** --username **redacted** --password **redacted** --store-password-in-clear-text"
DEBUG: exec completed (repository=**redacted** , branch=**redacted** )
       "cmd": "dotnet nuget update source **redacted** --username **redacted** --password **redacted** --store-password-in-clear-text",
       "durationMs": 442,
       "stdout": "Package source \"**redacted**\" was successfully updated.\n",
       "stderr": ""
DEBUG: Executing command (repository=**redacted** , branch=**redacted** )
       "command": "dotnet restore **redacted**  --force-evaluate"
DEBUG: exec completed (repository=**redacted** , branch=**redacted**)
       "cmd": "dotnet restore **redacted** --force-evaluate",
       "durationMs": 2422,
       "stdout": "  Determining projects to restore...\n  Restored **redacted** (in 947 ms).\n",
       "stderr": ""
DEBUG: Executing command (repository=**redacted**, branch=**redacted**)
       "command": "dotnet nuget update source **redacted** --username '' --password '' --store-password-in-clear-text"
DEBUG: Failed to generate lock file (repository=**redacted** , branch=**redacted** )
       "err": {
         "killed": false,
         "code": 1,
         "signal": null,
         "cmd": "dotnet nuget update source **redacted** --username '' --password '' --store-password-in-clear-text",
         "stdout": "",
         "stderr": "System.ArgumentException: Value cannot be null or whitespace. (Parameter 'token')\n   at Microsoft.DotNet.Cli.CommandLine.OptionError..ctor(String message, String token, AppliedOption option) in /_/src/source/CommandLine/OptionError.cs:line 21\n   at Microsoft.DotNet.Cli.CommandLine.Parser.UnrecognizedArg(String arg) in /_/src/source/CommandLine/Parser.cs:line 170\n   at System.Linq.Enumerable.SelectListIterator`2.MoveNext()\n   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)\n   at System.Collections.Generic.List`1.AddRange(IEnumerable`1 collection)\n   at Microsoft.DotNet.Cli.CommandLine.Parser.Parse(IReadOnlyCollection`1 rawArgs, Boolean isProgressive) in /_/src/source/CommandLine/Parser.cs:line 111\n   at Microsoft.DotNet.Cli.CommandLine.Parser.Parse(String[] args) in /_/src/source/CommandLine/Parser.cs:line 31\n   at Microsoft.DotNet.Cli.ParserExtensions.ParseFrom(Parser parser, String context, String[] args)\n   at Microsoft.DotNet.Cli.Program.ProcessArgs(String[] args, ITelemetry telemetryClient)\n   at Microsoft.DotNet.Cli.Program.Main(String[] args)\n",
         "message": "Command failed: dotnet nuget update source **redacted** --username '' --password '' --store-password-in-clear-text\nSystem.ArgumentException: Value cannot be null or whitespace. (Parameter 'token')\n   at Microsoft.DotNet.Cli.CommandLine.OptionError..ctor(String message, String token, AppliedOption option) in /_/src/source/CommandLine/OptionError.cs:line 21\n   at Microsoft.DotNet.Cli.CommandLine.Parser.UnrecognizedArg(String arg) in /_/src/source/CommandLine/Parser.cs:line 170\n   at System.Linq.Enumerable.SelectListIterator`2.MoveNext()\n   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)\n   at System.Collections.Generic.List`1.AddRange(IEnumerable`1 collection)\n   at Microsoft.DotNet.Cli.CommandLine.Parser.Parse(IReadOnlyCollection`1 rawArgs, Boolean isProgressive) in /_/src/source/CommandLine/Parser.cs:line 111\n   at Microsoft.DotNet.Cli.CommandLine.Parser.Parse(String[] args) in /_/src/source/CommandLine/Parser.cs:line 31\n   at Microsoft.DotNet.Cli.ParserExtensions.ParseFrom(Parser parser, String context, String[] args)\n   at Microsoft.DotNet.Cli.Program.ProcessArgs(String[] args, ITelemetry telemetryClient)\n   at Microsoft.DotNet.Cli.Program.Main(String[] args)\n",
         "stack": "Error: Command failed: dotnet nuget update source **redacted** --username '' --password '' --store-password-in-clear-text\nSystem.ArgumentException: Value cannot be null or whitespace. (Parameter 'token')\n   at Microsoft.DotNet.Cli.CommandLine.OptionError..ctor(String message, String token, AppliedOption option) in /_/src/source/CommandLine/OptionError.cs:line 21\n   at Microsoft.DotNet.Cli.CommandLine.Parser.UnrecognizedArg(String arg) in /_/src/source/CommandLine/Parser.cs:line 170\n   at System.Linq.Enumerable.SelectListIterator`2.MoveNext()\n   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)\n   at System.Collections.Generic.List`1.AddRange(IEnumerable`1 collection)\n   at Microsoft.DotNet.Cli.CommandLine.Parser.Parse(IReadOnlyCollection`1 rawArgs, Boolean isProgressive) in /_/src/source/CommandLine/Parser.cs:line 111\n   at Microsoft.DotNet.Cli.CommandLine.Parser.Parse(String[] args) in /_/src/source/CommandLine/Parser.cs:line 31\n   at Microsoft.DotNet.Cli.ParserExtensions.ParseFrom(Parser parser, String context, String[] args)\n   at Microsoft.DotNet.Cli.Program.ProcessArgs(String[] args, ITelemetry telemetryClient)\n   at Microsoft.DotNet.Cli.Program.Main(String[] args)\n\n    at ChildProcess.exithandler (child_process.js:308:12)\n    at ChildProcess.emit (events.js:314:20)\n    at ChildProcess.EventEmitter.emit (domain.js:483:12)\n    at maybeClose (internal/child_process.js:1021:16)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)"
       }

```

**To Reproduce**

1. Create a .NET Core project and reference an outdated package.
2. Add a `nuget.config` file defining a feed that requires username and password:

   ```xml
   <configuration>
     <packageSources>
       <clear />
       <add key="nuget.org"
          value="https://api.nuget.org/v3/index.json" />
       <add key="PrivateFeed"
          value="my-private-feed-url" />
     </packageSources>
   </configuration>
   ```

3. Set credentials in Renovate `hostRules` config:

   ```json
   {
     "hostRules": [
       {
         "timeout": 60000,
         "baseUrl": "my-private-feed-url",
         "hostType": "nuget",
         "username": "the-username",
         "password": "the-password"
       }
     ]
   }
   ```

I could not provide an example repo because I can't share credentials to a private feed...

**Additional context**

NuGet lock file support was added in #7375.

From a quick search in code, it seems the command with empty user and pass is run by [lib/manager/nuget/artifacts.ts:37](https://github.com/renovatebot/renovate/blob/master/lib/manager/nuget/artifacts.ts#L37).

### PR DESCRIPTION
**Pull request by fgreinacher**
## Changes:

We now create a temporary NuGet.config files to authenticate at private registries during lock file update. 

## Context:

Closes #7769 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added unit tests, or
- [ ] Unit tests + ran on a real repository

### CODE REVIEW

**Comment by fgreinacher** [comment]
@viceice @rarkins Any idea what's going wrong here: https://github.com/renovatebot/renovate/pull/7779/checks?check_run_id=1435648957#step:11:28?

File: lib/manager/nuget/artifacts.ts [hunk 1]
```diff
@@ -1,3 +1,6 @@
+import { join } from 'path';
+import cryptoRandomString from 'crypto-random-string';
+import { fs } from '../../../test/util';
```
Comment (viceice) [inline lib/manager/nuget/artifacts.ts:3]: Wrong import. 😉

Comment (fgreinacher) [reply lib/manager/nuget/artifacts.ts]: Argh, thanks!

**Review by viceice — CHANGES_REQUESTED** [review]
You need to add the folder as a volume for docker mode

**Comment by fgreinacher** [comment]
> You need to add the folder as a volume for docker mode

I [used `ensureCacheDir`](https://github.com/renovatebot/renovate/pull/7779/files#diff-983595d394c7e65d7e9afd380c9215a7aebfbd3d5e6f80fe8e31491dd4a7d72dR55) assuming the root cache dir would already be mounted. Not right?

**Review by viceice — APPROVED** [review]
You are right, cacheDir is already mounted.

**Pull request merged**